Buy/Graph.Task.swift

```
@@ -26,42 +26,271 @@
+        //  MARK: - Control -
         //
-        internal func setTask(_ task: URLSessionDataTask) {
-            self.task = task
+        func resume() {
+            self.isResumed = true
+            self.resume(using: self.cachePolicy)
+        }
+        
+        private func resume(using cachePolicy: CachePolicy) {
+            
+            let request    = self.request
+            let completion = self.completion
+            let hash       = request.hash
+            
+            switch cachePolicy {
+            case .cacheFirst(let expireIn):
+                print("Exercising cache policy: CACHE_FIRST(\(expireIn))")
+                
+                self.cachedModelFor(hash, expireIn: expireIn) { response in
+                    if let response = response {
+                        completion(response, nil)
+                        return
+                    } else {
+                        self.resume(using: .networkOnly)
+                    }
+                }
+                
+            case .cacheOnly:
+                print("Exercising cache policy: CACHE_ONLY")
+                
+                self.cachedModelFor(hash) { response in
+                    completion(response, nil)
+                    return
+                }
+                
+            case .networkFirst(let expireIn):
+                print("Exercising cache policy: NETWORK_FIRST(\(expireIn))")
+                
+                self.task = self.graphTaskWith(request, retryHandler: self.retryHandler) { response, data, error in
+                    
+                    if let _ = response, let data = data {
+                        self.cache(data, for: hash)
+                        completion(response, error)
+                        
+                    } else {
+                        
+                        self.cachedModelFor(hash, expireIn: expireIn) { response in
+                            if let response = response {
```

Because we have hit the cache only after a network error, which means the `error` is always not `nil` in this scope. We don't want to return the error if the cache has a response.